IAWG: First pass at publish_lookup chapter (#398)
pmix/pmix-standard: Chap_API_Publish.tex

\chapter{Publish/Lookup Operations}
\label{chap:pub}

Chapter~\ref{chap:api_rsvd_keys}
and
Section ~\ref{chap:data_sharing:non_rsvd_keys}
present how reserved and non-reserved keys deal with
information that either is associated with a specific process (i.e., the
retrieving process knows the identifier of the process that posted it) or
requires a synchronization operation prior to retrieval (e.g., the case of
globally unique non-reserved keys). However, another requirement exists for an
asynchronous exchange of data where neither the posting nor the retrieving

> process is known in advance, for example, two namespaces that do not share a child-parent relationship.

Reviewer: ```suggestion
process is known in advance (e.g., two namespaces that do not share a child-parent relationship).
```

The blocking form of this call will block until it has obtained confirmation from the datastore that the data is available for lookup. The \refarg{info} array can be released upon return from the blocking function call.

Publishing duplicate keys is permitted provided they are published to different

> ranges. Custom ranges are consider different if they have different members.

Reviewer 2: ```suggestion
ranges. Custom ranges are considered different if they have different members.
```

The blocking form of this call will block until it has obtained confirmation from the datastore that the data is available for lookup. The \refarg{info} array can be released upon return from the blocking function call.

Publishing duplicate keys is permitted provided they are published to different
ranges. Custom ranges are consider different if they have different members.
Duplicate keys being published on the same data range shall return the
\refconst{PMIX_ERR_DUPLICATE_KEY} error.

> In some cases, implementations may be incapable of distinguishing which

Reviewer: Note from 3Q 2022 ASC (Aug. 11, 2022):
 - Let's require `PMIX_DATA_TO_PUBLISH` as the only way to publish data.
   - This has some backward compatibility issues.
     - Open MPI and (maybe) MPICH would be impacted
   - We should add an error if `PMIX_DATA_TO_PUBLISH` is not provided in the publish call.
   - In v5, mark not using `PMIX_DATA_TO_PUBLISH` as deprecated.
 - Should we just create a new API with an additional required parameter for the data_to_publish (what to name it? `PMIx_Publish_data`?)
   - Then we just deprecate the old API instead of adding a new attribute to work around it.
```c
pmix_status_t PMIx_Publish_NEW(const pmix_info_t info_to_publish[], size_t ninfos,
                               const pmix_info_t info_directives[], size_t ndirs);
```

Reviewer: Per IAWG on Aug. 15, 2022:
 * Add new `PMIx_Publish2` and `pmix_server_publish2_fn_t` with two sets of `pmix_info_t`s
 * Deprecate the `PMIx_Publish` API and corresponding `pmix_server_publish_fn_t` API.
 * Do not add the `PMIX_DATA_TO_PUBLISH` attribute. Instead, prefer the new API.
 * In the Revision appendix note, `PMIx_Publish` was deprecated and replaced by `PMIx_Publish2` (same with server API)
 * No longer an issue on the ordering of info data and directives (this approach avoids this whole issue)

info keys in the \refarg{info} array are for publishing and which info keys are
directives.  To make it clear, it is recommended that the keys to be published
are designated by passing them as a \refstruct{pmix_data_array_t} using the
\refattr{PMIX_DATA_TO_PUBLISH} directive.
If the \refarg{info} array contains a \refattr{PMIX_DATA_TO_PUBLISH} info,
all other elements of the info array will be treated as directives.
If the info array does not include a \refattr{PMIX_DATA_TO_PUBLISH} info,
the implementation should
distinguish between info array elements that specify keys and directives as follows:
All standardized directives to the publish call,
including optional attributes the implementation does not support,
should be treated as
directives.  Non-supported directives
may be ignored as outlined in Section \ref{intro:portability:attributes},
but should not be treated as data to
publish.  The implementation may treat any custom (non-standardized) directives it
supports as directives.  All other \refarg{info} array elements

> should it be assumed to be data to be published.

Reviewer: ```suggestion
should be assumed to be data to be published.
```

The blocking form of this call will block until it has obtained confirmation from the datastore that the data is available for lookup. The \refarg{info} array can be released upon return from the blocking function call.

Publishing duplicate keys is permitted provided they are published to different
ranges. Custom ranges are consider different if they have different members.
Duplicate keys being published on the same data range shall return the
\refconst{PMIX_ERR_DUPLICATE_KEY} error.

> publishing to a \refconst{PMIX_RANGE_CUSTOM}

Reviewer: ```suggestion
Publishing to a \refconst{PMIX_RANGE_CUSTOM}
```

Author: now a nop

range which does not include the publisher will prevent

> any processes from using \refapi{PMIx_Lookup} to access the published data.

Reviewer 3: I'm actually not sure that statement is correct - the range determines who can access it. I'm not sure why the publisher would need to be a member of that group. The publisher always "owns" the data and can "unpublish" it.

> any processes from using \refapi{PMIx_Lookup} to access the published data.

Author: Good catch.  At one point in our revision process we had determined that the publisher and process calling Lookup both had to be part of the group, but we decided to remove that restriction and allow the publisher to not be part of the group of processes, so this should have been removed when we made that switch.

In some cases, implementations may be incapable of distinguishing which
info keys in the \refarg{info} array are for publishing and which info keys are
directives.  To make it clear, it is recommended that the keys to be published
are designated by passing them as a \refstruct{pmix_data_array_t} using the
\refattr{PMIX_DATA_TO_PUBLISH} directive.
If the \refarg{info} array contains a \refattr{PMIX_DATA_TO_PUBLISH} info,
all other elements of the info array will be treated as directives.
If the info array does not include a \refattr{PMIX_DATA_TO_PUBLISH} info,

> the implementation should
> distinguishing between info array elements that specify keys and directives as follows:

Reviewer: ```suggestion
the implementation should
distinguish between info array elements that specify keys and directives as follows:
```

Author: done

Undefined range.
%
\declareconstitemvalue{PMIX_RANGE_RM}{1}

> Published data and generated events are restricted to processes executing under the same instance of the host environment as the publisher or event creator.  Lookup of data is restricted to data published by processes running under the same instance of the host environment as the requester.

Reviewer: Note from 3Q 2022 (Aug. 11, 2022)
 * Old text was intended. That the published data is intended solely for the host environment not for application/tool processes.
   * Suggest leave the old text
 * `PMIX_RANGE_GLOBAL` : Anyone under that RM instance (basically what the new text is saying)

%
\declareconstitemvalue{PMIX_RANGE_SESSION}{4}
Published data and generated events are restricted to processes in the same session as the publisher or event creator.
Lookup of data is restricted to data published by procesess in the same session as the requester.
%
\declareconstitemvalue{PMIX_RANGE_GLOBAL}{5}
Published data and generated events are available to all processes within the domain of the host environment.
Lookup of data is unrestricted and open to data published by any processes within the domain of the host enivornment as the requester.  This range differs from \refconst{PMIX_RANGE_RM} only on systems which have mechanisms to share events and
publish/lookup data across multiple instances of a host environment.
%
\declareconstitemvalue{PMIX_RANGE_PROC_LOCAL}{7}
Published data and generated events are available only to calling process.
Lookup of data is restricted to data published by the calling process.
%
\declareconstitemvalue{PMIX_RANGE_CUSTOM}{6}
Published data and generated events are restricted to processes
described in the \refstruct{pmix_info_t} associated with this call.
Lookup of data is restricted to data published by the processes described in

> in the \refstruct{pmix_info_t}.

Reviewer: Note from 3Q 2022 (Aug. 11, 2022)
 - What does it mean for two `PMIX_RANGE_CUSTOM`s to match? Is it a strict ordering or is a subset fine?
 - Is the publisher required to be part of the range?
   - Publisher does need to be a member of the range?
   - "If I publish that data I own that data and can look it up so that I can unpublish"
     - Current description would prevent the lookup but not the unpublish.
     - Owner should always be able to look it up even if not in the range. Maybe need a clause here? What about overlapping keys?
 - Publish means "put some data for a set of consumers to lookup". The consumers might not know who published the data. For a custom range, those looking up might not know the full group. Alternatively, the lookup could specify a list including the publisher.

Reviewer: Per IAWG on Aug. 15, 2022:
 * Does the publisher have to be part of the range it publishes to?
   * The publisher does not need to be a member of the range.
   * The publisher is the owner of the data, so it is the only one able to unpublish the data, but must specify the range where that data was published.
   * The publisher is the owner of the data but cannot lookup the data if it is not part of the range in which it was pushed.
 * If I'm looking up from a `PMIX_RANGE_CUSTOM` does it have to match what the publisher provided?
   * `PMIX_RANGE_NAMESPECE` (example)
     * **Publish**: means that the data is to be available to a set of lookup'ers in my namespace.
     * **Lookup**: means that I want data provided by a publisher in my namespace.
   * `PMIX_RANGE_CUSTOM`
     * **Publish**: 'custom' means I want the data to be available to this set of lookup'ers. Might not include the publisher.
     * **Lookup**: 'custom' means I want data provided by one of these publishers. Does not need to know all of the processes that have access to the data - only who is providing the data.
   * Generally `PMIX_RANGE_` means:
     * **Publish**: Define who can look up the data
     * **Lookup**: Define who published the data
 * Does the range have to match on the publish and lookup?
   * I'm publishing to range X, and I'm looking up on range X.
   * `PMIX_RANGE_CUSTOM` is the only one that takes an argument list of processes. It's the only one where the publisher and lookup'er might be using the same value associated with the `PMIX_RANGE_CUSTOM` key.
 * What if the publisher publishes the same key (different values) to different custom ranges that both include the client looking up the key? Or two different publishers both publish the same key (different values) that include the lookup client in different `PMIX_RANGE_CUSTOM`s.
   * This is not allowed.
   * The Server should reject the lookup operation as conflicting entries. The client should then change their custom range to include one
 * What If I want to lookup a value published by a process outside of my namespace?
   * Publisher will use `PMIX_RANGE_SESSION` or `PMIX_RANGE_CUSTOM` (if they know the lookup'er)
   * Lookup'er will user `PMIX_RANGE_CUSTOM` with the publisher's name
 * **Future work**: Maybe extend `PMIX_RANGE_NAMESPACE` with an argument of which namespace. Currently, it is the same as the caller.

Undefined range.
%
\declareconstitemvalue{PMIX_RANGE_RM}{1}
Published data and generated events are restricted to processes executing under the same instance of the host environment as the publisher or event creator.  Lookup of data is restricted to data published by processes running under the same instance of the host environment as the requester.
%

> \declareconstitemvalue{PMIX_RANGE_LOCAL}{2}

Reviewer 4: Reading this text I stumbled again on the question of wether ranges are "nested" or not. The new text in `PMIx_Lookup` below makes it very clear that it is not, but some forewarning here (or a ref to the lookup rules  `\ref{chap:pub:retrules}.`) could help clarify.

Reviewer: Per IAWG on Aug. 15, 2022:
 * We should make clean that though the ranges are overlapping in definition, they are not nested. So ranges must match.
 * We should add the forward reference in the top paragraph of this section to the lookup section where this is more clearly noted.

\endgroup

The retrieval rules for published data primarily revolve around enforcing data access permissions and range constraints.
All publish and lookup operations operate on a range. If not specified, the range defaults to \refconst{PMIX_RANGE_SESSION}.
The key being looked up will match with a published key only if all of the following conditions are met:

\begin{enumerate}
    \item The lookup key matches the published key.
    \item The publisher of the published key falls within the lookup range specified by the requester.

>     \item The the requester falls within the range specified by the publisher.

Author: got it.  Will show up in next push.

    \item If the publisher specified access permissions, the effective \ac{UID} and \ac{GID} of the requester must meet those requirements.
\end{enumerate}

> Fig.~\ref{fig:publish_lookup} shows process P0 executing on Host X in Namespace 1 in Session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.

Author: I agree.  I was just trying to capture what I thought the document says currently, but I agree it is very complex to understand and to implement and it still has non-determinant cases (results that may vary based on implementation).   I'm all for making something that is as simple as possible both to understand and implement, even if that means it is not very rich in features.  Like you said, this is just meant to be a simple rendezvous mechanism.   I need to think more about what options we have and what we could do without breaking backward compatibility.  This is one of those cases where I don't think anyone is invested in the way things "work" now, but not sure how much liberty we should take on changing behavior.  

On another note, I would love to see the Publish API take 2 info arrays... one for the things to publish and one for modifiers to the call.  As it is now if I call Publish with an array of infos like:  

PMIX_RANGE=PMIX_RANGE_NAMESPACE
tree=oak
PMIX_RANGE=PMIX_RANGE_GLOBAL
tree=elm
PMIX_TIMEOUT=10

then we have several problems:  If the implementation doesn't handle or recognize PMIX_TIMEOUT then it will actually publish PMIX_TIMEOUT as a key since it doesn't recognize it as a directive.  Also, will it publish tree=oak at the namespace range and tree=elm at the Global range or will it publish only one of the values of tree at some range (either namespace or global)?   Things get really tricky when the things to publish are combined with the directives.

But deprecating existing API's for new Publish API's seems like a lot of work for something that very few products are using and probably have figured out a way to not get tripped up by these things already.  I don't want the pmix-standard to be a burden to implementations or progress, but I would like it to describe how things actually work so someone can develop new clients without getting lost or confused.   I'll see if I can think up some way to simplify what is supported without breaking backward compatibility.

> Fig.~\ref{fig:publish_lookup} shows process P0 executing on Host X in Namespace 1 in Session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.

Reviewer 3: So don't deprecate the API, but add an attribute for specifying modifiers to the call. It would take a `pmix_data_array_t` containing an array of `pmix_info_t` that has the modifiers in them. I don't think that would be much of a problem for backwards compatibility.

If you prefer, you could add an attribute for the things to publish, leaving all else in the call to be modifiers. Either way should be fine.

I wouldn't worry too much about changes in behavior at this point. OMPI is the primary user of it (for connect/accept), and I've already outlined a way for them to use `PMIx_Group` instead - which is a far better and more performant solution for that use-case.

> Fig.~\ref{fig:publish_lookup} shows process P0 executing on Host X in Namespace 1 in Session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.

Author: I like that idea.   We could have both approaches for a while, then move towards deprecating the old way after a while.

> Fig.~\ref{fig:publish_lookup} shows process P0 executing on Host X in Namespace 1 in Session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.

Reviewer 3: Thinking some more, we should probably add the attribute for the things to publish (as opposed to the modifiers) so we remain consistent with the other APIs - whose info arrays all contain modifiers.

because the only publisher is not within P1's specified range which covers only its
local host.  Similarly, P1 will not successfully lookup a value for "tree" on the range
\refconst{PMIX_RANGE_NAMESPACE} since the only publisher is not within P1's namespace.
A lookup of "tree" by P1 on the range \refconst{PMIX_RANGE_GLOBAL} does permit possible matches of keys published by P0.  P0 has published "tree" three times,
but P1 is outside of the P0's node and namespace ranges.  Therefore, only the value of "tree"
published to the session level is retrievable by P1.  Notice that the lookup range
does not need to match the publishing range for a match to occur.  The requirement
is that the requester is in the range specified by the publisher and the publisher is in the range specified by the requester.

The figure shows a similar scenario for P2.  However, unlike P1, P2 is in the same namespace as P0 which results in two possible published values of
"tree" that could be returned by a call to \refapi{PMIx_Lookup} by P2 on the range \refconst{PMIX_RANGE_SESSION} or \refconst{PMIX_RANGE_NAMESPACE}.

> An implementation is required to return one of the possible matches in such cases where

Reviewer 3: Yeah, I think we have to preclude this possibility - it should be an error.

Key, process identifier, and data will all be copied into the destination \refstruct{pmix_pdata_t} - thus, the source \refstruct{pmix_pdata_t} may free'd without affecting the copied data once the macro has completed.
\adviceuserend

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Retrieval rules for published data}
\label{chap:pub:retrules}

The retrieval rules for published data primarily revolve around enforcing data access permissions and range constraints.
All publish and lookup operations operate on a range. If not specified, the range defaults to \refconst{PMIX_RANGE_SESSION}.
The key being looked up will match with a published key only if all of the following conditions are met:

\begin{enumerate}
    \item The lookup key matches the published key.

>     \item The type of range specified by the publisher is the same as the type of range specified by the requester.

Reviewer: Is this required (e.g., requestor specifies 'node' range, and publisher specifies 'global' range)? Or is it enough that the requestor be part of the publisher's range?

Author: The new model is that the ranges must match... if you publish to 'global', then the requestor must specifically lookup on 'global' range.

>     \item If a custom range is specified by the publisher and the requester, the members described in both cases must be identical.  The publisher is not required to be a member of a custom range.

Reviewer 4: This seems to contradict the rule from https://github.com/pmix/pmix-standard/pull/398#discussion_r940372741

Author: Yes, this is correct.  The other location is wrong.

>     \item The requestor must be a member of the publisher's range.

Reviewer: ```suggestion
    \item The requestor must be a member of the range specified by the publisher.
```
